assets/app/features/upgrade/steps.factory.js

```diff
@@ -95,5 +101,138 @@
+            var stepToStateMap = {
+                    'upgrade_prechecks': 'upgrade-landing',
+                    'upgrade_prepare': 'upgrade-landing',
+                    'admin_backup': 'upgrade.backup',
+                    'admin_repo_checks': 'upgrade.administration-repository-checks',
+                    'admin_upgrade': 'upgrade.upgrade-administration-server',
+                    'database': 'upgrade.database-configuration',
+                    'nodes_repo_checks': 'upgrade.nodes-repositories-checks',
+                    'nodes_services': 'upgrade.openstack-services',
+                    'nodes_db_dump': 'upgrade.openstack-services',
+                    'nodes_upgrade': 'upgrade.upgrade-nodes',
+                    'finished': 'upgrade.upgrade-nodes',
+                },
+                currentStep = statusData.current_step,
+                steps = statusData.steps,
+                currentState = stepToStateMap[currentStep];
+
+            // select previous step if current is still pending and previous step has a different page
+            if (steps[currentStep].status === STEP_STATES.pending) {
+                var storedLastState = localStorage.getItem(UPGRADE_LAST_STATE_KEY);
+
+                // if there is some stored state and it's pointing to page for current pending step
+                // show this instead of the previous one, otherwise apply the logic below to find
+                // proper page to show
+                if (storedLastState === currentState) {
+                    return currentState;
+                }
+
+                var backendStepsWithCurrentState = keysForValue(stepToStateMap, currentState);
+                // if there are more "backend" steps matching currentState, only move back if currentStep
+                // is the first one on the list
+                if (backendStepsWithCurrentState.length > 1 && backendStepsWithCurrentState[0] !== currentStep) {
+                    return currentState;
+                }
+
+                var currentStepData = stepByState(currentState);
+
+                // getting `undefined` from `stepByState()` should happen only
+                // for landing page which is not included in the wizard UI.
+                if (angular.isDefined(currentStepData)) {
+                    if (currentStepData.id > 0) {
+                        return stepByID(currentStepData.id - 1).state;
+                    } else {
+                        // don't move user back from backup page to landing page
+                        // this would require another "prepare" call to move forward
+                    }
+                } else if (currentState === 'upgrade-landing'){
+                    // this is the only valid case when expectedStep can be undefined
+                    // expectedState is already on landing page so nothing needs to be done
+                    // keeping this empty block here to make logic clear
```

Comment:
A comment at the beginning of the conditional section, or even at the end explaining the remaining cases should be enough and will also help to reduce the complexity of this method